Add Sqlsrv driver

michalsn opened the pull request:
**Description**
This PR adds Sqlsrv driver based on @qury work. I'm not experienced with MSSQL but not many changes were needed.

The code has been cleaned up a bit and a few minor changes have been made to make the tests working.

Some tweaks are probably needed, so if anyone takes the time to review it, I would appreciate adding code suggestions to make things easier for me.

Closes #3612

**Checklist:**
- [x] Securely signed commits
- [x] Unit testing, with >80% coverage
- [x] Conforms to style guide

Commit:
Update Sqlsrv driver

Commit:
implement _limit() using row ... fetch

Commit:
fix delete with limit

Commit:
add case for Sqlsrv for  PreparedQueryTest

Commit:
fix PreparedQuery

Commit:
add Sqlsrv to DbUtilsTest and ForgeTest

Commit:
add Sqlsrv to DBUtilsTest

Commit:
add Sqlsrv to ForgeTest

Commit:
add Sqlsrv to ForgeTest

Commit:
Add null coalescing operator to insertID() and begin adding  support for schemas

Commit:
Update formatting and continue adding schema support

Commit:
Update testListTables to use DBPrefix

Commit:
Allow dropping columns that are part of an index by droping the index first.

Commit:
Update formatting

Commit:
Update formatting

qury committed:
Update formatting

qury committed:
Update formatting

qury committed:
fix dataSeek

qury committed:
Fix Unique key identification

qury committed:
Implement replace()

qury committed:
Add db and schema to droptable

qury committed:
Move drop table under assert.

qury committed:
Fix _droptable typo

qury committed:
Fix testCompositeKey in ForgeTest

qury committed:
Add groupby fields to select
if select portion is not speficied in the builder.
Fixes model behaviour.

qury committed:
Repalce array spread with array_merge
to preserve PHP 7.2 compatibility

qury committed:
fix ForgeTest again

qury committed:
for ForgeTest again

qury committed:
Update Database Type test table for Sqlsrv

qury committed:
Fix issue with testPaginateForQueryWithGroupBy

qury committed:
Remove id from insert into an autoid field.

qury committed:
Remove id from insert statement
as it is clashing with an autoid field.

qury committed:
Fix testSaveUpdateRecordObject

qury committed:
Fix testUpdateNoPrimaryKey test

qury committed:
fix compileSelect

michalsn committed:
Add Sqlsrv driver with tests

lonnieezell commented:
You rock! Thanks for getting this one across the finish line, and @qury - you're a rockstar! Thanks for all of your work on this.

michalsn committed:
Make rector happy

mostafakhudair commented on .github/workflows/test-phpunit.yml:
```diff
@@ -19,7 +19,7 @@ jobs:
       matrix:
         php-versions: ['7.2', '7.3', '7.4']
-        db-platforms: ['MySQLi', 'Postgre', 'SQLite3']
+        db-platforms: ['MySQLi', 'Postgre', 'SQLite3', 'Sqlsrv']
```
```suggestion
        db-platforms: ['MySQLi', 'Postgre', 'SQLite3', 'SQLSRV']
```

mostafakhudair commented on system/Database/Sqlsrv/Builder.php:
```diff
@@ -0,0 +1,691 @@
+ * @filesource
+ */
+
+namespace CodeIgniter\Database\Sqlsrv;
```
```suggestion
namespace CodeIgniter\Database\SQLSRV;
```

mostafakhudair commented on system/Database/Sqlsrv/Connection.php:
```diff
@@ -0,0 +1,625 @@
+ * @filesource
+ */
+
+namespace CodeIgniter\Database\Sqlsrv;
```
```suggestion
namespace CodeIgniter\Database\SQLSRV;
```

mostafakhudair commented on system/Database/Sqlsrv/Connection.php:
```diff
@@ -0,0 +1,625 @@
+	 *
+	 * @var string
+	 */
+	public $DBDriver = 'SQLSrv';
```
```suggestion
	public $DBDriver = 'SQLSRV';
```

mostafakhudair commented on system/Database/Sqlsrv/Forge.php:
```diff
@@ -0,0 +1,444 @@
+ * @filesource
+ */
+
+namespace CodeIgniter\Database\Sqlsrv;
```
```suggestion
namespace CodeIgniter\Database\SQLSRV;
```

mostafakhudair commented on system/Database/Sqlsrv/PreparedQuery.php:
```diff
@@ -0,0 +1,157 @@
+ * @filesource
+ */
+
+namespace CodeIgniter\Database\Sqlsrv;
```
```suggestion
namespace CodeIgniter\Database\SQLSRV;
```

mostafakhudair commented on system/Database/Sqlsrv/Result.php:
```diff
@@ -0,0 +1,212 @@
+ * @filesource
+ */
+
+namespace CodeIgniter\Database\Sqlsrv;
```
```suggestion
namespace CodeIgniter\Database\SQLSRV;
```

mostafakhudair commented on system/Database/Sqlsrv/Utils.php:
```diff
@@ -0,0 +1,79 @@
+ * @filesource
+ */
+
+namespace CodeIgniter\Database\Sqlsrv;
```
```suggestion
namespace CodeIgniter\Database\SQLSRV;
```

mostafakhudair commented on tests/_support/Config/Registrar.php:
```diff
@@ -71,6 +71,25 @@ class Registrar
+			'username' => 'sa',
+			'password' => '1Secure*Password1',
+			'database' => 'test',
+			'DBDriver' => 'Sqlsrv',
```
```suggestion
			'DBDriver' => 'SQLSRV',
```

mostafakhudair commented on tests/_support/Database/Migrations/20160428212500_Create_test_tables.php:
```diff
@@ -93,119 +93,137 @@ public function up()
+			);
+		}
+
+		if ($this->db->DBDriver === 'Sqlsrv')
```
```suggestion
		if ($this->db->DBDriver === 'SQLSRV')
```

mostafakhudair commented on tests/_support/Database/Migrations/20160428212500_Create_test_tables.php:
```diff
@@ -93,119 +93,137 @@ public function up()
+			unset($data_type_fields['type_timestamp']);
 		}
+
+		if ($this->db->DBDriver === 'Postgre' || $this->db->DBDriver === 'Sqlsrv')
```
```suggestion
		if ($this->db->DBDriver === 'Postgre' || $this->db->DBDriver === 'SQLSRV')
```

mostafakhudair commented on tests/_support/Database/Seeds/CITestSeeder.php:
```diff
@@ -115,6 +117,20 @@ public function run()
 			unset($data['type_test'][0]['type_blob']);
 		}
+
+		if ($this->db->DBDriver === 'Sqlsrv')
```
```suggestion
		if ($this->db->DBDriver === 'SQLSRV')
```

mostafakhudair commented on tests/system/Database/Live/DbUtilsTest.php:
```diff
@@ -58,13 +59,7 @@ public function testUtilsListDatabases()
-			$this->assertTrue(in_array('test', $databases));
-		}
-		elseif ($this->db->DBDriver === 'Postgre')
+		if (in_array($this->db->DBDriver, ['MySQLi', 'Postgre', 'Sqlsrv']))
```
```suggestion
		if (in_array($this->db->DBDriver, ['MySQLi', 'Postgre', 'SQLSRV']))
```

mostafakhudair commented on tests/system/Database/Live/DbUtilsTest.php:
```diff
@@ -85,13 +80,7 @@ public function testUtilsDatabaseExist()
-			$this->assertTrue($exist);
-		}
-		elseif ($this->db->DBDriver === 'Postgre')
+		if (in_array($this->db->DBDriver, ['MySQLi', 'Postgre', 'Sqlsrv']))
```
```suggestion
		if (in_array($this->db->DBDriver, ['MySQLi', 'Postgre', 'SQLSRV']))
```

mostafakhudair commented on tests/system/Database/Live/DbUtilsTest.php:
```diff
@@ -139,13 +128,13 @@ public function testUtilsOptimizeTable()
 		$d = $util->optimizeTable('db_job');
 
-		if ($this->db->DBDriver === 'Postgre' || $this->db->DBDriver === 'SQLite3')
+		if (in_array($this->db->DBDriver, ['SQLite3', 'Postgre', 'Sqlsrv']))
```
```suggestion
		if (in_array($this->db->DBDriver, ['SQLite3', 'Postgre', 'SQLSRV']))
```

mostafakhudair commented on tests/system/Database/Live/ForgeTest.php:
```diff
@@ -7,11 +9,11 @@
  */
-class ForgeTest extends CIDatabaseTestCase
-{
+class ForgeTest extends CIDatabaseTestCase {
```
```suggestion
class ForgeTest extends CIDatabaseTestCase
{
```

mostafakhudair commented on tests/system/Database/Live/ForgeTest.php:
```diff
@@ -606,6 +612,16 @@ public function testAddFields()
 
 			$this->assertEquals($fieldsData[1]->default, null);
 		}
+		elseif ($this->db->DBDriver === 'Sqlsrv')
```
```suggestion
		elseif ($this->db->DBDriver === 'SQLSRV')
```

mostafakhudair commented on tests/system/Database/Live/GetTest.php:
```diff
@@ -162,6 +162,23 @@ public function testGetFieldData()
 			$this->assertEquals('timestamp', $type_test[11]->type_name);  //TIMESTAMP
 			$this->assertEquals('int8', $type_test[12]->type_name); //BIGINT
 		}
+		if ($this->db->DBDriver === 'Sqlsrv')
```
```suggestion
		if ($this->db->DBDriver === 'SQLSRV')
```

mostafakhudair commented on tests/system/Database/Live/MetadataTest.php:
```diff
@@ -1,15 +1,16 @@
-class MetadataTest extends CIDatabaseTestCase
-{
-	protected $refresh = true;
+class MetadataTest extends CIDatabaseTestCase {
```
```suggestion
class MetadataTest extends CIDatabaseTestCase
{
```

mostafakhudair commented on tests/system/Database/Live/ModelTest.php:
```diff
@@ -478,8 +478,14 @@ public function testSaveUpdateRecordObject()
+		$data = new \stdClass();
+
+		// Sqlsrv does not allow forcing an ID into an autoincrement field.
+		if ($this->db->DBDriver !== 'Sqlsrv')
```
```suggestion
		if ($this->db->DBDriver !== 'SQLSRV')
```

mostafakhudair commented on tests/system/Database/Live/PreparedQueryTest.php:
```diff
@@ -7,11 +9,10 @@
-class PreparedQueryTest extends CIDatabaseTestCase
-{
-	protected $refresh = true;
+class PreparedQueryTest extends CIDatabaseTestCase {
```
```suggestion
class PreparedQueryTest extends CIDatabaseTestCase
{
```

mostafakhudair commented on tests/system/Database/Migrations/MigrationRunnerTest.php:
```diff
@@ -76,8 +76,18 @@ public function testGetHistory()
 			'batch'     => 1,
 		];
 
+		if ($this->db->DBDriver === 'Sqlsrv')
```
```suggestion
		if ($this->db->DBDriver === 'SQLSRV')
```

mostafakhudair commented on tests/system/Database/Migrations/MigrationRunnerTest.php:
```diff
@@ -76,8 +76,18 @@ public function testGetHistory()
+
 		$this->hasInDatabase('migrations', $history);
 
+		if ($this->db->DBDriver === 'Sqlsrv')
```
```suggestion
		if ($this->db->DBDriver === 'SQLSRV')
```

mostafakhudair commented on .github/workflows/test-phpunit.yml:
```diff
@@ -39,22 +39,34 @@ jobs:
 
     steps:
+      - name: Create database for MSSQL Server
+        if: ${{ matrix.db-platforms == 'Sqlsrv'}}
```
```suggestion
        if: ${{ matrix.db-platforms == 'SQLSRV'}}
```

mostafakhudair commented on system/Database/Sqlsrv/Connection.php:
```diff
@@ -0,0 +1,625 @@
+	 */
+	public function getError()
+	{
+		$error        = [
```
```suggestion
		$error = [
```

mostafakhudair commented on system/Database/Sqlsrv/Connection.php:
```diff
@@ -0,0 +1,625 @@
+			'code'    => '00000',
+			'message' => '',
+		];
+		$sqlsrvErrors = sqlsrv_errors(SQLSRV_ERR_ERRORS);
```
```suggestion

		$sqlsrvErrors = sqlsrv_errors(SQLSRV_ERR_ERRORS);
```

mostafakhudair commented on tests/_support/Config/Registrar.php:
```diff
@@ -71,6 +71,25 @@ class Registrar
 			'failover' => [],
 			'port'     => 3306,
 		],
+		'Sqlsrv'  => [
```
```suggestion
		'SQLSRV'  => [
```

mostafakhudair commented on tests/system/Database/Live/ForgeTest.php:
```diff
@@ -165,6 +167,10 @@ public function testCreateTableApplyBigInt()
 		{
 			$this->assertEquals(strtolower($fieldsData[0]->type), 'integer');
 		}
+		elseif ($this->db->DBDriver === 'Sqlsrv')
```
```suggestion
		elseif ($this->db->DBDriver === 'SQLSRV')
```

mostafakhudair commented on tests/system/Database/Live/ForgeTest.php:
```diff
@@ -678,6 +694,20 @@ public function testCompositeKey()
 			$this->assertEquals($keys['db_forge_test_1_code_active']->name, 'db_forge_test_1_code_active');
 			$this->assertEquals($keys['db_forge_test_1_code_active']->fields, ['code', 'active']);
 		}
+		elseif ($this->db->DBDriver === 'Sqlsrv')
```
```suggestion
		elseif ($this->db->DBDriver === 'SQLSRV')
```

qury replied:
Any reason to uppercase SQLSRV?

mostafakhudair replied:
The Driver name was mentioned in uppercase in php manual https://www.php.net/manual/en/book.sqlsrv.php also in Microsoft website reference https://docs.microsoft.com/en-us/sql/connect/php/sqlsrv-driver-api-reference?view=sql-server-ver15
so I think it would be nice to keep the driver name style in uppercase also its easier to read just like other drivers MySQL, SQLite3, PostgreSQL, etc..

michalsn replied:
Personally, I don't think that this is worth changing. We already have other drivers that obviously don't follow the name convention from PHP. From what I know other frameworks also have their own names for database drivers. As long as there is no confusion about what driver it is, we probably should stay with the current name.

michalsn committed:
Apply suggestions from code review

Co-authored-by: Mostafa Khudair <59371810+mostafakhudair@users.noreply.github.com>

samsonasik commented on system/Database/Sqlsrv/Builder.php:
```diff
@@ -0,0 +1,691 @@
+		{
+			return '"' . $this->db->getDatabase() . '"."' . $this->db->schema . '"."' . str_replace('"', '', $table) . '"';
+		}
+		else
```
no need else for return early, the check can be handled by rector at https://github.com/codeigniter4/CodeIgniter4/pull/3715

michalsn replied:
I was removing these things from the code but the new rule for Rector will be very useful.

samsonasik commented:
Documentation `user_guide_src/source/intro/requirements.rst` need to be updated  below: 

```
A database is required for most web application programming.
Currently supported databases are:
```

and remove `MS SQL` below : 

```
Not all of the drivers have been converted/rewritten for CodeIgniter4.
The list below shows the outstanding ones.
```

michalsn committed:
User guide update [ci skip]

samsonasik commented:
You can fix rector error with run : 

```bash
vendor/bin/rector process
```

michalsn committed:
Rector fixes

michalsn committed:
Try to speed up the tests

michalsn committed:
Try to speed up the tests

michalsn committed:
Try to speed up the tests

michalsn committed:
Try to speed up the tests

michalsn committed:
Code style fixes

michalsn committed:
Code style fixes

michalsn committed:
Code style fixes

michalsn committed:
Code style fixes

michalsn committed:
Stop printing errors

michalsn committed:
Stop printing errors

michalsn committed:
Stop printing errors

michalsn committed:
Stop printing errors

MGatner commented:
I'm sorry I haven't had a chance to look at this yet. I've only used MSSQL moderately but I did follow the original work. I suspect this is good but it is a pretty big addition and merits a full review.

Is there anyone else watching this who is familiar with MSSQL and could do a review?

samsonasik commented:
I tried in ci4-album with Sqlsrv driver, with album - track table relation at https://github.com/samsonasik/ci4-album/pull/8/checks?check_run_id=1209486951 , running test got the following error: 

```bash
   ┐
   ├ Exception: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Cannot truncate table 'album' because it is being referenced by a FOREIGN KEY constraint.
   │
   ╵ /home/runner/work/ci4-album/ci4-album/vendor/codeigniter4/codeigniter4/system/Database/Sqlsrv/Connection.php:539
   ╵ /home/runner/work/ci4-album/ci4-album/vendor/codeigniter4/codeigniter4/system/Database/BaseConnection.php:724
   ╵ /home/runner/work/ci4-album/ci4-album/vendor/codeigniter4/codeigniter4/system/Database/BaseConnection.php:652
   ╵ /home/runner/work/ci4-album/ci4-album/vendor/codeigniter4/codeigniter4/system/Database/BaseBuilder.php:2774
   ╵ /home/runner/work/ci4-album/ci4-album/test/Controller/AlbumTest.php:25
   ┴
```

The foreign key is using :

```
$this->forge->addForeignKey('album_id', 'album', 'id', 'CASCADE', 'CASCADE');
```

at https://github.com/samsonasik/ci4-album/blob/ac5d31f53beb67384d82b90bbe2e0da2d80492d0/src/Database/Migrations/2020-03-26-203957_Track.php#L29

which should automatically handle truncating like in MySQL.

qury commented:
The database itself does not allow truncating a table that has a remote constraint.
The only solution would be to drop the constraint, truncate the table and re-create the constraint.

samsonasik commented:
ok, another issue I found is it somehow can't do `model->db->affectedRows()` got undefined property stdClass::$numrows

michalsn commented:
I believe that problem with `TRUNCATE` will also occur in Postgre, so if we want to add this functionality (truncate when foreign keys are present), we should probably address this in the separate PR that will add support to all handlers (via special parameter for this method).

As for `numrows`, I will try to look into it.

michalsn committed:
Changes for comparison query
 - this needed some changes because SQL-Type TEXT is not compatible for comparison with strings

michalsn committed:
compileSelect update

michalsn commented:
@samsonasik It should be fixed now.

paulbalandan commented on system/Database/Sqlsrv/Builder.php:
```diff
@@ -0,0 +1,688 @@
+use CodeIgniter\Database\Exceptions\DataException;
+use CodeIgniter\Database\ResultInterface;
+
+/**
+ * Builder for Sqlsrv
+ */
+class Builder extends BaseBuilder
+{
+	/// TODO: auto check for TextCastToInt
+	/// TODO: auto check for InsertIndexValue
+	/// TODO: replace: delete index entries before insert
```
Can we use proper `@todo` markings here so that IDEs and PHPDoc API can see that these are TO-DO for this class? Something like:
```suggestion
/**
 * Builder for Sqlsrv
 *
 * @todo auto check for TextCastToInt
 * @todo auto check for InsertIndexValue
 * @todo replace: delete index entries before insert
 */
class Builder extends BaseBuilder
{
```

paulbalandan commented on system/Database/Sqlsrv/Builder.php:
```diff
@@ -0,0 +1,688 @@
+	}
+
+	/**
+	 * Get funn name of the table
```
```suggestion
	 * Get full name of the table
```

samsonasik commented:
@michalsn thank you. The `stdClass::$numrows` error seems fixed now. 

The next seems find a tweak how to make controller test that extends `CIDatabaseTestCase` works without error : "Cannot truncate table 'album' because it is being referenced by a FOREIGN KEY constraint"

michalsn committed:
Apply suggestions from code review

Co-authored-by: John Paul E. Balandan, CPA <51850998+paulbalandan@users.noreply.github.com>

michalsn commented:
@samsonasik Is there no issue with `truncate` on Postgre?

samsonasik commented:
I tried, but not in run tests, I will check.

qury commented:
@michalsn 
If a table1 has a foreign key to table 2 then you can truncate table1 without any issues however you can't truncate table2.


SQL used
`ALTER TABLE [dbo].[table1]
ADD FOREIGN KEY ([table2]) REFERENCES [dbo].[table2] ([id]) ON DELETE CASCADE ON UPDATE CASCADE`

samsonasik commented:
@michalsn @qury Thank you. Just tried run test in PostgreSQL ,and it got same error: 

```bash
 ✘ Index album has no data
   ┐
   ├ ErrorException: pg_query(): Query failed: ERROR:  cannot truncate a table referenced in a foreign key constraint
   ├ DETAIL:  Table "track" references "album".                                                                      
   ├ HINT:  Truncate table "track" at the same time, or use TRUNCATE ... CASCADE. 
```

It seems that need manual clean up data in controller test for Postgres and Sqlsrv then.

michalsn commented:
@qury I was referring to the code example given by @samsonasik and he is using `truncate()` combined with `disableForeignKeyChecks()` ([link](https://github.com/samsonasik/ci4-album/blob/ac5d31f53beb67384d82b90bbe2e0da2d80492d0/test/Controller/AlbumTest.php#L24)) which will work for MySQL and SQLite (probably). 

That's why I proposed to address this in the separate PR to make it possible to produce the same functionality for all drivers. Probably something like `truncate(true)`? idk really - I haven't thought it out yet. And I'm not sure if this will happen.

Anyway for now, `truncate` "problem" is out of the scope for this PR, since this is completely valid behavior.

michalsn commented:
@qury We have some commented lines in the code. Is it something that will be useful when doing some changes later or it can be cleared up?

qury commented:
@michalsn They can be cleared up. 
Also thanks for taking this over :+1:

michalsn committed:
Cleanup inline comments [ci skip]

michalsn committed:
Rector cleanup

michalsn committed:
Fix orderBy() test

michalsn committed:
Entity support for result object

michalsn commented:
Seems like we have a temporary problem with `sqlsrv` extension (not available) after updating from PHP 7.4.10 to PHP 7.4.11. I tried to use a specific version number (the latest preview) as a workaround but without a luck 😕

samsonasik commented:
is it possible to pin php-version to 7.4.10 ?

michalsn commented:
No, we can only specify a minor version 😞

michalsn commented:
Well... apparently on Monday everything decided to start working again 🤷‍♂️

michalsn commented:
@lonnieezell @MGatner @samsonasik @paulbalandan @qury All tests look good. I think I would like to merge it at this stage. This will allow other developers to battle-test it even more and then we should be able to fix any possible bugs before the next release.

MGatner commented:
I agree. I still haven't actually used it but I've been through the code and read all the comments, and I will be very glad to see this get some use!

lonnieezell reviewed (approved):
Agreed. I've been trying to follow the thread, and it sounds like you've got most of the kinks worked out. Huge thanks for taking this one across the finish line!

michalsn commented:
One more time - thank you @qury!

robsonpiere commented:
Good work!

Is anyone working on the version with PDO ? 
I'd like to help.

I have used the driver made available by microsoft (with CI3) and has served me very well.

https://github.com/microsoft/msphpsql/releases
